### PR TITLE
Build and release DLite-Python

### LINKED ISSUE #123: Set up automatical deployment
Dlite can now create Deploy Python wheel, deb and rpm packages. Set up a github workflow that automatically generate and deploys these.

Python wheel should be deployed on pypi
deb and rpm packages could be deployed by attaching them to the github release

Ideally the deb and rpm packages should just use libpython.so and not be picky about the exact version of python the user has installed on his/her system.

### LINKED ISSUE #173: Deploy dlite on pypi
Create a dlite-python pypi project and set up automatic deployment to it.

### LINKED ISSUE #178: No respect for virtual environment system paths
It seems dlite does not respect the lib paths within a virtual environment, forcing loading from `~/.local` on UNIX systems instead of retrieving where the virtual environment has it's `lib`, `bin`, `share`, etc. folders.

I've tried amending the `python/setup.py:CMAKE_ARGS` variable for Linux systems without luck.

Steps to reproduce:
- Setup a [`virtualenv`](https://virtualenv.pypa.io/en/latest/) virtual environment using an appropriate Python version, e.g., 3.7.
- Add either or all of the following entries to `CMAKE_ARGS` in the `"Linux"` block of `python/setup.py`:
  - `f"-DPython3_EXECUTABLE={sys.executable}"`
  - `f"-DCMAKE_INSTALL_PREFIX={sys.prefix}"`
  - `"-DPython3_FIND_VIRTUALENV=ONLY"`
- Run `cd python && pip install .`.
  Everything should install nicely, if not, make sure to update `pip`, `setuptools`, `wheel` and install `numpy`.
- Install `pandas` and `tables`: `pip install pandas tables`.
- Run the read CSV example: `cd ../examples/read-csv && python readcsv.py`

The last step fails, complaining the "'yaml' api cannot be found".
It then references the `~/.local/...` folders where it's trying to find the `yaml` driver, and failing.

Expected behaviour:
After installation the read CSV example should work.

### LINKED ISSUE #209: pip installing from GitHub fails for Python3.9 on Windows
When pip installing from GitHub, i.e.

`pip install dlite-python@git+https://github.com/SINTEF/dlite.git@1d7b4ccec0e76799a25992534cd295a80d83878a#subdirectory=python`

an error is raised

### PR DESCRIPTION
**Pull request by CasperWA**
Use the [PyPA manylinux images](https://github.com/pypa/manylinux) to build and subsequently release DLite-Python on PyPI.
Use [PyPA cibuildwheel](https://cibuildwheel.readthedocs.io) to build several wheels.

Closes #173 
Closes #123 
Fixes #178 
Closes #209 

---

Still to do:
- [X] Set up infrastructure to use `cibuildwheel`.
- [X] Set up Linux Docker images and release/store in GitHub Container Registry.
- [x] Move CI job for building Docker images to a separate cron-job (nightly) build.
- [x] Add checking step for building Docker images about whether the `musllinux`/`manylinux` `latest` tag-images have changed before building new images. (Create separate issue+PR for implementing this.)
- [x] Refine matrix of different builds in `cibuildwheel` CI job.
  - [X] `manylinux2014_x86_64`
  - [X] `manylinux2014_i686` (commented out, i.e., not being built)
  - [x] `manylinux2010_x86_64`
  - [x] `manylinux2010_i686`
  - [x] `musllinux_1_1_x86_64` (will postpone for separate PR)
  - [x] `musllinux_1_1_i686`
  - [x] `win_amd64`
  - [x] `win32` (will postpone for separate PR)
  - [x] `macosx_x86_64` (will postpone for separate PR)
- [x] Add subsequent wheel tests for `cibuildwheel` CI jobs.
- [x] Move CI job for building wheels with `cibuildwheel` to a release workflow.
- [x] Add publish to PyPI step to `cibuildwheel` job.

### CODE REVIEW

**Comment by CasperWA** [comment]
In the last commits I have removed several workflow files in favor of new ones.
I think it'd be wise to go through the existing/old workflow files and determine whether or not they can be updated/removed.

**Comment by CasperWA** [comment]
@jesper-friis should fix the failing tests in `musllinux_1_1_x86_64 before this is completely ready for review.

**Comment by jesper-friis** [comment]
I am not able to reproduce the failure manually: https://github.com/SINTEF/dlite/runs/5369470376?check_suite_focus=true

I have tried the following:
* Build the docker image using your script in .github/docker/run_cibuildwheel.sh
* Run the docker image with: ` `
* Build musllinux x86_64 with 
   ```shell
   /opt/python/cp310-cp310/bin/cmake \
     -DWITH_DOC=OFF \
     -DWITH_JSON=ON \
     -DWITH_HDF5=OFF \
     -DALLOW_WARNINGS=ON \
     -Ddlite_PYTHON_BUILD_REDISTRIBUTABLE_PACKAGE=YES \
     -DPython3_FIND_VIRTUALENV=ONLY \
     -DPython3_FIND_IMPLEMENTATIONS=CPython \
     -DPython3_EXECUTABLE=/opt/python/cp310-cp310/bin/python3.10 \
     ..
   make
   ```
* Running the test with `ctest` - works fine

How do I reproduce what cibuildwheel does inside the docker image?

**Comment by CasperWA** [comment]
> How do I reproduce what cibuildwheel does inside the docker image?

Excellent question. I'm afraid I don't have the answer :/

Is `ctest` running the same tests that we are with `test_python_bindings.py`?
Can you try to create a virtual environment within the docker image then after building DLite, and then installing DLite-Python in that virtual environment and run the tests there? Because that's what `cibuidwheel` is doing, I believe. It's creating a virtual environment, installing the generated wheel and running the commands we gave it to run `test_python_bindings.py`.

**Comment by CasperWA** [comment]
Let's get this in now - and setup an issue for the failing build to be solved at a later point.

**Comment by CasperWA** [comment]
Issues generated from this PR: #220, #221, and #225

**Comment by CasperWA** [comment]
@jesper-friis and @pscff this is ready for a review.

File: python/setup.py [hunk 1]
```diff
@@ -88,111 +123,103 @@ def build_extension(self, ext):
...
             raise
 
         cmake_bdist_dir = Path(self.build_temp) / Path(ext.python_package_dir)
         dir_util.copy_tree(
             str(cmake_bdist_dir / ext.name), str(Path(output_dir) / ext.name)
         )
 
-
-requirements = [
+extra_requirements = [
```
Comment (CasperWA) [inline python/setup.py:159]: Make hard requirements of these, until the "index"-idea has been implemented (or something similar).

File: src/tests/python/CMakeLists.txt [hunk 2]
```diff
@@ -42,8 +42,8 @@ foreach(test ${tests})
       ENVIRONMENT "PYTHONHOME=${Python3_RUNTIME_LIBRARY_DIRS}")
   endif()
   # No need to set PYTHONPATH since we are not importing the package in any test here
-  #set_property(TEST ${test} APPEND PROPERTY
-  #  ENVIRONMENT "PYTHONPATH=${dlite_PYTHONPATH_NATIVE}")
+  set_property(TEST ${test} APPEND PROPERTY
```
Comment (CasperWA) [inline src/tests/python/CMakeLists.txt:45]: @jesper-friis Should this be commented out again?

File: storages/json/tests/CMakeLists.txt [hunk 3]
```diff
@@ -39,7 +39,17 @@ endif()
 
 foreach(test ${tests})
   add_executable(${test} ${test}.c)
-  target_link_libraries(${test} dlite-plugins-json)
+  target_link_libraries(${test}
+    dlite
+    dlite-plugins-json
+    ${Python3_STATIC_LIBS}
+    )
+  #target_link_options(${test} PRIVATE
```
Comment (CasperWA) [inline storages/json/tests/CMakeLists.txt:47]: @jesper-friis clean up

File: storages/python/CMakeLists.txt [hunk 4]
```diff
@@ -60,5 +59,9 @@ install(
   )
 
 # tests
-add_subdirectory(tests-c)
+# For now don't compile json tests if linking with static Python
```
Comment (CasperWA) [inline storages/python/CMakeLists.txt:62]: Make an issue for this @jesper-friis

File: src/tests/python/CMakeLists.txt (hunk 2 above)
Comment (jesper-friis) [reply src/tests/python/CMakeLists.txt]: yes, I think so

**Pull request merged**